Source: RDKM (aishwarya_natarajan2@comcast.com)
Index: kernel-source/include/linux/netfilter/x_tables.h
===================================================================
--- kernel-source.orig/include/linux/netfilter/x_tables.h
+++ kernel-source/include/linux/netfilter/x_tables.h
+@@ -36,6 +36,7 @@
  		const void *matchinfo, *targinfo;
  	};
  	const struct nf_hook_state *state;
-+	unsigned int 	hooknum;
- 	int fragoff;
++	unsigned int    hooknum;
  	unsigned int thoff;
+ 	u16 fragoff;
  	bool hotdrop;
Index: kernel-source/include/linux/timer.h
===================================================================
--- kernel-source.orig/include/linux/timer.h
+++ kernel-source/include/linux/timer.h
@@ -15,6 +15,7 @@ struct timer_list {
 	 */
 	struct hlist_node	entry;
 	unsigned long		expires;
+	unsigned long            data;
 	void			(*function)(struct timer_list *);
 	u32			flags;
 
Index: kernel-source/include/net/netfilter/nf_conntrack.h
===================================================================
--- kernel-source.orig/include/net/netfilter/nf_conntrack.h
+++ kernel-source/include/net/netfilter/nf_conntrack.h
@@ -24,6 +24,8 @@
 
 #include <net/netfilter/nf_conntrack_tuple.h>
 
+#define NF_CT_ASSERT(x)
+
 struct nf_ct_udp {
 	unsigned long	stream_ts;
 };
Index: kernel-source/net/ipv4/netfilter/ipt_TRIGGER.c
===================================================================
--- kernel-source.orig/net/ipv4/netfilter/ipt_TRIGGER.c
+++ kernel-source/net/ipv4/netfilter/ipt_TRIGGER.c
@@ -79,7 +79,7 @@ static void trigger_refresh(struct ipt_t
 {
 	DEBUGP( KERN_LEVL "%s: \n", __FUNCTION__);
 	NF_CT_ASSERT(trig);
-    spin_lock_bh(&nf_conntrack_lock);
+    spin_lock_bh(&nf_conntrack_expect_lock);
 
 	/* Need del_timer for race avoidance (may already be dying). */
 	if (del_timer(&trig->timeout)) {
@@ -87,7 +87,7 @@ static void trigger_refresh(struct ipt_t
 		add_timer(&trig->timeout);
 	}
 
-    spin_unlock_bh(&nf_conntrack_lock);
+    spin_unlock_bh(&nf_conntrack_expect_lock);
 }
 
 static void __del_trigger(struct ipt_trigger *trig)
@@ -100,14 +100,14 @@ static void __del_trigger(struct ipt_tri
 	kfree(trig);
 }
 
-static void trigger_timeout(unsigned long ul_trig)
+static void trigger_timeout(struct timer_list *t)
 {
-	struct ipt_trigger *trig= (void *) ul_trig;
+	struct ipt_trigger *trig= from_timer(trig, t,timeout);
 
 	DEBUGP( KERN_LEVL "trigger list %p timed out\n", trig);
-	spin_lock_bh(&nf_conntrack_lock);
+	spin_lock_bh(&nf_conntrack_expect_lock);
 	__del_trigger(trig);
-    spin_unlock_bh(&nf_conntrack_lock);
+    spin_unlock_bh(&nf_conntrack_expect_lock);
 }
 
 static unsigned int
@@ -116,12 +116,12 @@ add_new_trigger(struct ipt_trigger *trig
 	struct ipt_trigger *new;
 
 	DEBUGP( KERN_LEVL "!!!!!!!!!!!! %s !!!!!!!!!!!\n", __FUNCTION__);
-	spin_lock_bh(&nf_conntrack_lock);
+	spin_lock_bh(&nf_conntrack_expect_lock);
 	new = (struct ipt_trigger *)
 		kmalloc(sizeof(struct ipt_trigger), GFP_ATOMIC);
 
 	if (!new) {
-		spin_unlock_bh(&nf_conntrack_lock);
+		spin_unlock_bh(&nf_conntrack_expect_lock);
 		DEBUGP( KERN_LEVL "%s: OOM allocating trigger list\n", __FUNCTION__);
 		return -ENOMEM;
 	}
@@ -133,13 +133,13 @@ add_new_trigger(struct ipt_trigger *trig
 	/* add to global table of trigger */
 	list_add(&trigger_list, &new->list);
 	/* add and start timer if required */
-	init_timer(&new->timeout);
+	//init_timer(&new->timeout);
 	new->timeout.data = (unsigned long)new;
 	new->timeout.function = trigger_timeout;
 	new->timeout.expires = jiffies + (TRIGGER_TIMEOUT * HZ);
-	add_timer(&new->timeout);
+	timer_setup(&new->timeout,trigger_timeout,0);
 		   
-    spin_unlock_bh(&nf_conntrack_lock);
+    spin_unlock_bh(&nf_conntrack_expect_lock);
 
     return 0;
 }
@@ -304,7 +304,7 @@ trigger_in(struct sk_buff *skb)
     return XT_CONTINUE;	/* the match was done but no trigger exist  */
 }
 
-static void xt_nat_convert_range(struct nf_nat_range *dst,
+static void xt_nat_convert_range(struct nf_nat_range2 *dst,
 				 const struct nf_nat_ipv4_range *src)
 {
 	memset(&dst->min_addr, 0, sizeof(dst->min_addr));
@@ -334,6 +334,7 @@ trigger_dnat(struct sk_buff *skb, int ho
 	enum ip_conntrack_info ctinfo;
 	struct nf_nat_ipv4_multi_range_compat newrange;
     struct nf_nat_range range;
+    struct nf_nat_range2 range_info;
 
 	DEBUGP( KERN_LEVL "####Starting  %s ############\n", __FUNCTION__);
     DEBUGP( KERN_LEVL "TRIGGER_DNAT: Protocol ->%s  SRC_IP->%pI4"
@@ -368,7 +369,7 @@ trigger_dnat(struct sk_buff *skb, int ho
 	newrange.range[0].min.all = 0;
 	newrange.range[0].min.all = 0;
 
-	xt_nat_convert_range(&range, &newrange.range[0]);
+	xt_nat_convert_range(&range_info, &newrange.range[0]);
 
         /* 
 	*  We call here to create the nat processor  that will replace packet source address 
@@ -385,7 +386,7 @@ trigger_dnat(struct sk_buff *skb, int ho
         */
 	
 	/* Alter the destination of incoming packet. */
-	return nf_nat_setup_info(ct, &range, NF_NAT_MANIP_DST);
+	return nf_nat_setup_info(ct, &range_info, NF_NAT_MANIP_DST);
 }
 
 
